ledger-service/http-json/src/main/scala/com/digitalasset/http/dbbackend/SupportedJdbcDriver.scala

```diff
@@ -3,53 +3,85 @@
 
 package com.daml.http.dbbackend
 
-/** A nullary typeclass of incompatible JDBC operations and settings, selected by
+import scalaz.Liskov, Liskov.<~<
+
+/** Incompatible JDBC operations and settings, selected by
   * the options passed to json-api at startup.
-  *
-  * We currently use the fact that it is nonsensical to run json-api against more
-  * than one JDBC driver ''in the image'' as justification to make this typeclass
-  * nullary.  If that changes in the future, a phantom type parameter should be
-  * introduced so as to distinguish instances by type.
   */
-sealed abstract class SupportedJdbcDriver private (
+final class SupportedJdbcDriver[+Q] private (
     label: String,
     private[http] val retrySqlStates: Set[String],
+    private[http] val q: Q,
 ) {
-  type SqlInterpol
-  private[http] val queries: Queries.Aux[SqlInterpol]
-  private[http] implicit val ipol: SqlInterpol
-  override def toString = s"SupportedJdbcDriver($label)"
+  import SupportedJdbcDriver.{Staged, Configured}
+  def configure(tablePrefix: String, extraConf: Map[String, String])(implicit
+      Q: Q <~< Staged
+  ): Either[String, SupportedJdbcDriver.TC] = {
```

Comment:
> would you mind explaining these things to me

No, but it would be much simpler if you read [this article](https://typelevel.org/blog/2014/07/02/type_equality_to_leibniz.html) first. `<~<` is just a slightly tricky variation on that.

Reply:
OK will read that article. If questions are left I will come back to you then :smile:

Reply:
I think I roughly understood what is going on here. Interesting code.